QccPack-0.61-1/src/libQccPackIMG/utils/man1/imgdist.1
.TH IMGDIST 1 "QCCPACK" ""
.SH NAME
imgdist \- calculate the distortion between two images
.SH SYNOPSIS
.B imgdist
.RB "[\|" \-vo "\|]"
.RB "[\|" \-mse "\|]"
.RB "[\|" \-snr "\|]"
.RB "[\|" \-psnr "\|]"
.RB "[\|" \-mae "\|]"
.RB "[\|" \-cie "\|]"
.RB "[\|" \-m
.IR  mask "\|]"
.I imgfile1
.I imgfile2
.SH OPTIONS
.TP
.B \-vo
Output only the value of the distortion (terse output).
.TP
.B \-mse
Express the distortion in terms of MSE.
.TP
.B \-snr
Express the distortion in terms of SNR.
.TP
.B \-psnr
Express the distortion in terms of PSNR.
.TP
.B \-mae
Express the distortion in terms of MAE.
.TP
.B \-cie
Express the distortion in terms of an SNR based
on the 1964 CIE color distortion measure
in the CIE Modified UCS color space.  Requires that
.I imgfile1
and
.I imgfile2
be color images.
.TP
.BI \-m " mask"
String. Name of file containing the transparency mask for
shape-adaptive distortion.
.SH DESCRIPTION
.LP
.B imgdist
prints
the mean square error (MSE), signal-to-noise ratio (SNR),
peak signal-to-noise ratio (PSNR),
maximum absolute error (MAE),
or CIE-based color SNR 
between two image files,
.I imgfile1
and
.IR imgfile2 ,
to stdout.
.LP
The options
.BR \-mse ,
.BR \-snr ,
.BR \-psnr ,
.BR \-mae ,
and
.B \-cie
indicate in which form the distortion should be expressed.
If none of these
options are specified, 
.B imgdist
prints the distortion in all forms.
.LP
If the option
.B \-vo
is specified,
then 
.B imgdist
prints a terse output, consisting of only the value of the distortion,
to stdout.  One of the options
.BR \-mse ,
.BR \-snr ,
.BR \-psnr ,
.BR \-mae ,
or
.B \-cie
must be specified in this case.
.LP
For the calculation of SNR and PSNR,
.B imgfile1
is considered to be the original signal; 
.B imgdist
thus finds the variance of
.B imgfile1
for use in the SNR calculation.
For the calculation of PSNR,
.B imgdist
assumes that the peak-to-peak range of
.B imgfile1
is 255.
.LP
If the image files are color,
the MSE, SNR, PSNR, and MAE
distortion calculations are performed separately for the Y, U, and V 
components of the image.  The distortions for each image component are
output in order.
Additionally, for color images, a color SNR distortion based on
the 1964 CIE formula for the distance between two colors
is calculated
(see
.BR QccIMGImageColorSNR (3)).
.LP
The distortion calculation can be made "shape-adaptive"
by specification of the
.B \-m 
option, in which case
.I mask
gives the name of an image file that indicates the
transparency mask for the image.
.I mask
must be a grayscale image
in which black indicates transparency, white
indicates full opacity, and gray tones are degrees of
translucence
(see
.BR QccAlpha (3)).
This transparency mask indicates which regions in
.I imgfile1
and 
.I imgfile2
are to be taken into consideration in the distortion calculation---pixels
that are marked as transparent by
.I mask
are ignored and are not included in either the
calculation of pixel differences or in the total count of pixels as used
in the averaging of the differences.
.I mask
must be the same size as both
.I imgfile1
and
.IR imgfile2 .
Note: The CIE-based color SNR distortion cannot be calculated when a mask
is specified.
.SH "NOTES"
Jain defines the SNR in terms of the
log of the ratio of the signal variance to the MSE. However, Clarke
indicates that, historically, the SNR has not been defined consistently,
and describes slightly different variants, including the
variance-based definition used by Jain, as well as the log of the ratio of
the mean squared signal to the MSE, a definition popular with some
other authors. Clarke cautions that one must be sure of the definition
of the measure when comparing SNRs
as the various definitions yield different values.
.B imgdist
follows Jain's definition of a variance-based SNR.
.LP
Similarly, there are inconsistent definitions of the PSNR. The most logical
definition involves the log of the square of the actual peak-to-peak range
of the reference image to the MSE; this is the definition used by Jain.
However, the PSNR is now nearly universally accepted as the log of the
square of the
.I maximum possible
peak-to-peak range allowed by the pixels,
a value that depends on the bit-depth of the image pixels and not
the reference image itself. For example, an
8-bit image has a possible peak-to-peak range of 255, although the
actual range of any given image may be less than this.
.B imgdist
calculates PSNR using a peak-to-peak range of 255.
.SH "SEE ALSO"
.BR QccIMGImageComponentMSE (3),
.BR QccIMGImageComponentMAE (3),
.BR QccIMGImageComponentShapeAdaptiveMSE (3),
.BR QccIMGImageComponentShapeAdaptiveMAE (3),
.BR QccIMGImageColorSNR (3),
.BR QccPackIMG (3),
.BR QccPack (3)

Anil K. Jain, "Fundamentals of Digital Image Processing."
Englewood Cliffs, NJ: Prentice Hall, 1989.  (See Chapter 3).

Roger J. Clarke, "Digital Compression of Still Images and Video."
London: Academic Press Ltd., 1995.  (see Section 1.2.7).

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackIMG/utils/imgdist.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "imgdist.h"

#define USG_STRING "[-vo %:] [-mse %:] [-snr %:] [-psnr %:] [-mae %:] [-cie %:] [-m %: %s:mask] %s:imgfile1 %s:imgfile2"

#define ALL 0
#define MSE 1
#define SNR 2
#define PSNR 3
#define MAE 4
#define CIE 5

int value_only_flag = 0;
int mse_specified = 0;
int snr_specified = 0;
int psnr_specified = 0;
int mae_specified = 0;
int cie_specified = 0;

int DistortionMethod = ALL;

int ColorImages = 0;

QccIMGImage Image1;
QccIMGImage Image2;
int MaxVal1 = 255;
int NumCols1, NumRows1;
int NumCols2, NumRows2;

QccIMGImage Mask;
int MaskSpecified = 0;
int MaskNumRows, MaskNumCols;
int ColorMask;


int main(int argc, char *argv[])
{
  double value = 0;
  double mse = -1.0;
  double snr = -1.0;
  double psnr = -1.0;
  double mae = -1.0;
  double signal_power;
  QccIMGImageComponent *image_component1 = NULL;
  QccIMGImageComponent *image_component2 = NULL;
  QccIMGImageComponent *mask = NULL;
  int component, num_components;
  QccString component_name;
  int num_rows1_Y, num_cols1_Y;
  int num_rows1_U, num_cols1_U;
  int num_rows1_V, num_cols1_V;
  int num_rows2_Y, num_cols2_Y;
  int num_rows2_U, num_cols2_U;
  int num_rows2_V, num_cols2_V;
  int mask_num_rows_Y, mask_num_cols_Y;
  int mask_num_rows_U, mask_num_cols_U;
  int mask_num_rows_V, mask_num_cols_V;
  double cie_snr;

  QccInit(argc, argv);

  QccIMGImageInitialize(&Image1);
  QccIMGImageInitialize(&Image2);
  QccIMGImageInitialize(&Mask);
  
  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         &value_only_flag,
                         &mse_specified,
                         &snr_specified,
                         &psnr_specified,
                         &mae_specified,
                         &cie_specified,
                         &MaskSpecified,
                         Mask.filename,
                         Image1.filename, Image2.filename))
    QccErrorExit();

  if (mse_specified + snr_specified + psnr_specified +
      mae_specified + cie_specified > 1)
    {
      QccErrorAddMessage("%s: Only one of -mse, -snr, -psnr, -mae, and -cie can be specified",
                         argv[0]);
      QccErrorExit();
    }

  if (mse_specified)
    DistortionMethod = MSE;
  else if (snr_specified)
    DistortionMethod = SNR;
  else if (psnr_specified)
    DistortionMethod = PSNR;
  else if (mae_specified)
    DistortionMethod = MAE;
  else if (cie_specified)
    DistortionMethod = CIE;
  else
    DistortionMethod = ALL;

  if (value_only_flag &&
      (!mse_specified && !snr_specified && !psnr_specified &&
       !mae_specified && !cie_specified))
    {
      QccErrorAddMessage("%s: One of -mse, -snr, -psnr, -mae, and -cie must be used with -vo",
                         argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageRead(&Image1))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
                         argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageRead(&Image2))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccIMGImageGetSizeYUV(&Image1,
                            &num_rows1_Y, &num_cols1_Y,
                            &num_rows1_U, &num_cols1_U,
                            &num_rows1_V, &num_cols1_V))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSizeYUV()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccIMGImageGetSizeYUV(&Image2,
                            &num_rows2_Y, &num_cols2_Y,
                            &num_rows2_U, &num_cols2_U,
                            &num_rows2_V, &num_cols2_V))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSizeYUV()",
                         argv[0]);
      QccErrorExit();
    }

  if (MaskSpecified)
    {
      if (QccIMGImageRead(&Mask))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageGetSizeYUV(&Mask,
                                &mask_num_rows_Y, &mask_num_cols_Y,
                                &mask_num_rows_U, &mask_num_cols_U,
                                &mask_num_rows_V, &mask_num_cols_V))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageGetSizeYUV()",
                             argv[0]);
          QccErrorExit();
        }
      ColorMask = QccIMGImageColor(&Mask);
    }
  
  ColorImages = QccIMGImageColor(&Image1);
  if (ColorImages != QccIMGImageColor(&Image2))
    {
      QccErrorAddMessage("%s: Both images must be color or both must be grayscale",
                         argv[0]);
      QccErrorExit();
    }

  if (cie_specified && !ColorImages)
    {
      QccErrorAddMessage("%s: -cie valid only for color images",
                         argv[0]);
      QccErrorExit();
    }

  num_components = ColorImages ? 3 : 1;

  if (!value_only_flag)
    printf("The distortion between files\n   %s\n         and\n   %s\nis:\n",
           Image1.filename, Image2.filename);
  
  if (((DistortionMethod == CIE) || (DistortionMethod == ALL)) &&
      (ColorImages))
    {
      if (MaskSpecified)
        {
          QccErrorAddMessage("%s: CIE distortion unavailable when mask is specified",
                             argv[0]);
          QccErrorExit();
        }

      cie_snr = QccIMGImageColorSNR(&Image1, &Image2);
      
      if (!value_only_flag)
        {
          printf(" SNR based on 1964 CIE color-metric formula:\n");
          printf("%16.6f dB\n", cie_snr);
        }
      else
        printf("%f\n", cie_snr);

      if (cie_specified)
        QccExit;
    }

  for (component = 0; component < num_components; component++)
    {
      switch (component)
        {
        case 0:
          image_component1 = &(Image1.Y);
          image_component2 = &(Image2.Y);
          QccStringCopy(component_name, "Y component");
          NumRows1 = num_rows1_Y;
          NumCols1 = num_cols1_Y;
          NumRows2 = num_rows2_Y;
          NumCols2 = num_cols2_Y;
          if (MaskSpecified)
            {
              MaskNumRows = mask_num_rows_Y;
              MaskNumCols = mask_num_cols_Y;
              mask = &(Mask.Y);
            }
          break;
        case 1:
          image_component1 = &(Image1.U);
          image_component2 = &(Image2.U);
          QccStringCopy(component_name, "U component");
          NumRows1 = num_rows1_U;
          NumCols1 = num_cols1_U;
          NumRows2 = num_rows2_U;
          NumCols2 = num_cols2_U;
          if (MaskSpecified)
            {
              MaskNumRows = (ColorMask ? mask_num_rows_U : mask_num_rows_Y);
              MaskNumCols = (ColorMask ? mask_num_cols_U : mask_num_cols_Y);
              mask = (ColorMask ? &(Mask.U) : &(Mask.Y));
            }
          break;
        case 2:
          image_component1 = &(Image1.V);
          image_component2 = &(Image2.V);
          QccStringCopy(component_name, "V component");
          NumRows1 = num_rows1_V;
          NumCols1 = num_cols1_V;
          NumRows2 = num_rows2_V;
          NumCols2 = num_cols2_V;
          if (MaskSpecified)
            {
              MaskNumRows = (ColorMask ? mask_num_rows_V : mask_num_rows_Y);
              MaskNumCols = (ColorMask ? mask_num_cols_V : mask_num_cols_Y);
              mask = (ColorMask ? &(Mask.V) : &(Mask.Y));
            }
          break;
        }          

      if ((NumCols1 != NumCols2) || (NumRows1 != NumRows2))
        {
          QccErrorAddMessage("%s: Size of %s (%d x %d) does not equal that of %s (%d x %d) for image component %d",
                             argv[0],
                             Image1.filename, NumCols1, NumRows1,
                             Image2.filename, NumCols2, NumRows2,
                             component);
          QccErrorExit();
        }

      if (!MaskSpecified)
        {
          mse = QccIMGImageComponentMSE(image_component1, image_component2);
          mae = QccIMGImageComponentMAE(image_component1, image_component2);
          signal_power = QccIMGImageComponentVariance(image_component1);
        }
      else
        {
          if ((MaskNumRows != NumRows1) || (MaskNumCols != NumCols1))
            {
              QccErrorAddMessage("%s: Mask must be same size as images",
                                 argv[0]);
              QccErrorExit();
            }

          mse = QccIMGImageComponentShapeAdaptiveMSE(image_component1,
                                                     image_component2,
                                                     mask);

          mae = QccIMGImageComponentShapeAdaptiveMAE(image_component1,
                                                     image_component2,
                                                     mask);
          signal_power =
            QccIMGImageComponentShapeAdaptiveVariance(image_component1,
                                                      mask);
        }
      
      if (mse != 0.0)
        {
          psnr = 10.0*log10((double)MaxVal1*MaxVal1/mse);
          snr = 10.0*log10(signal_power/mse);
        }
      
      if (!value_only_flag)
        {
          if (num_components != 1)
            printf(" %s:\n", component_name);
          switch (DistortionMethod)
            {
            case MSE:
              printf("%16.6f (MSE)\n", mse);
              break;
            case SNR:
              printf("%16.6f dB (SNR)\n", snr);
              break;
            case PSNR:
              printf("%16.6f dB (PSNR)\n", psnr);
              break;
            case MAE:
              printf("%16.6f (MAE)\n", mae);
              break;
            case ALL:
            default:
              printf("%16.6f\t(MSE)\n%16.6f dB\t(SNR)\n%16.6f dB\t(PSNR)\n%16.6f\t(MAE)\n",
                     mse, snr, psnr, mae);
              break;
            }
        }
      else
        {
          switch (DistortionMethod)
            {
            case MSE:
              value = mse;
              break;
            case SNR:
              value = snr;
              break;
            case PSNR:
              value = psnr;
              break;
            case MAE:
              value = mae;
              break;
            }
          printf("%f\n", value);
        }
    }

  QccIMGImageFree(&Image1);
  QccIMGImageFree(&Image2);
  QccIMGImageFree(&Mask);

  QccExit;
}
